### PR TITLE
fix: salesforce v2 refactor

### PR DESCRIPTION
**Pull request by shrouti1507**
## What are the changes introduced in this PR?

salesforceResponseHandler to re-visit using switch case instead of multiple if-else statements

To have single logic of to infer from event for OAUTH or LEGACY implementation - https://github.com/rudderlabs/rudder-transformer/pull/2873#discussion_r1415313844 

## What is the related Linear task?

Resolves INT-1129

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [x] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [x] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [x] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

File: src/v0/destinations/salesforce/utils.js [hunk 1]
```diff
@@ -28,72 +25,74 @@ const ACCESS_TOKEN_CACHE = new Cache(ACCESS_TOKEN_CACHE_TTL);
...
-  if (!isHttpStatusSuccess(status) && status >= 400) {
-    const matchErrorCode = (errorCode) =>
-      response && Array.isArray(response) && response.some((resp) => resp?.errorCode === errorCode);
-    if (status === 401 && authKey && matchErrorCode('INVALID_SESSION_ID')) {
-      if (authorizationFlow === OAUTH) {
+  const matchErrorCode = (errorCode) =>
+    response && Array.isArray(response) && response.some((resp) => resp?.errorCode === errorCode);
+
+  const getErrorMessage = () => {
+    if (response && Array.isArray(response) && response[0]?.message?.length > 0) {
```
Comment (koladilip) [inline src/v0/destinations/salesforce/utils.js:32]: add comment about this condition, explaining what is this checking

File: src/v0/destinations/salesforce/utils.js [hunk 2]
```diff
@@ -28,72 +25,74 @@ const ACCESS_TOKEN_CACHE = new Cache(ACCESS_TOKEN_CACHE_TTL);
 const salesforceResponseHandler = (destResponse, sourceMessage, authKey, authorizationFlow) => {
   const { status, response } = destResponse;
 
-  // if the response from destination is not a success case build an explicit error
-  if (!isHttpStatusSuccess(status) && status >= 400) {
-    const matchErrorCode = (errorCode) =>
-      response && Array.isArray(response) && response.some((resp) => resp?.errorCode === errorCode);
-    if (status === 401 && authKey && matchErrorCode('INVALID_SESSION_ID')) {
-      if (authorizationFlow === OAUTH) {
+  const matchErrorCode = (errorCode) =>
```
Comment (koladilip) [inline src/v0/destinations/salesforce/utils.js:28]: attach documentation link for response structure

**Review by koladilip — COMMENTED** [review]
Add description to PR

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3773/test-report.html

File: src/v0/destinations/salesforce/utils.js (hunk 1 above)
Comment (shrouti1507) [reply src/v0/destinations/salesforce/utils.js]: done

File: src/v0/destinations/salesforce/utils.js (hunk 2 above)
Comment (shrouti1507) [reply src/v0/destinations/salesforce/utils.js:106]: done

File: src/v0/destinations/salesforce/utils.js [hunk 3]
```diff
@@ -47,59 +59,108 @@ const salesforceResponseHandler = (destResponse, sourceMessage, authKey, authori
...
+        handleAuthError(
+          'INVALID_SESSION_ID',
+          authKey,
+          authorizationFlow,
+          sourceMessage,
+          destResponse,
+          status,
+        );
+      }
+      handleAuthError('DEFAULT', authKey, authorizationFlow, sourceMessage, destResponse, status);
```
Comment (sanpj2292) [inline src/v0/destinations/salesforce/utils.js:123]:
```suggestion
      let errorCode = 'DEFAULT';
      if (authKey && matchErrorCode('INVALID_SESSION_ID')) {
        errorCode = 'INVALID_SESSION_ID';
      }
      handleAuthError(errorCode, authKey, authorizationFlow, sourceMessage, destResponse, status);
```

File: src/v0/destinations/salesforce/utils.js [hunk 4]
```diff
@@ -47,59 +59,108 @@ const salesforceResponseHandler = (destResponse, sourceMessage, authKey, authori
...
+      break;
+    case 403:
+      if (matchErrorCode('REQUEST_LIMIT_EXCEEDED')) {
+        throw new ThrottledError(
+          `${DESTINATION} Request Failed - due to "REQUEST_LIMIT_EXCEEDED", (Throttled) ${sourceMessage}`,
+          destResponse,
+        );
+      }
```
Comment (sanpj2292) [inline src/v0/destinations/salesforce/utils.js:131]:
```suggestion
      if (matchErrorCode('REQUEST_LIMIT_EXCEEDED')) {
        throw new ThrottledError(
          `${DESTINATION} Request Failed - due to "REQUEST_LIMIT_EXCEEDED", (Throttled) ${sourceMessage}`,
          destResponse,
        );
      }
      handleCommonAbortableError(destResponse, sourceMessage, status);
```
Should we change to this ?

File: src/v0/destinations/salesforce/utils.js [hunk 5]
```diff
@@ -47,59 +59,108 @@ const salesforceResponseHandler = (destResponse, sourceMessage, authKey, authori
...
+      break;
+
+    case 503:
+    case 500:
       throw new RetryableError(
-        `${DESTINATION} Request Failed - "${response.message}", (Retryable) ${sourceMessage}`,
+        `${DESTINATION} Request Failed - due to "${getErrorMessage(response)}", (Retryable) ${sourceMessage}`,
         500,
         destResponse,
       );
```
Comment (sanpj2292) [inline src/v0/destinations/salesforce/utils.js:155]:
```javascript
throw new RetryableError(
        `${DESTINATION} Request Failed - due to "${getErrorMessage(response)}", (Retryable) ${sourceMessage}`,
        status,
        destResponse,
      );
```
It is probably better to do this ?

Comment (shrouti1507) [reply src/v0/destinations/salesforce/utils.js]: done

File: src/v0/destinations/salesforce/utils.js [hunk 6]
```diff
@@ -47,59 +59,104 @@ const salesforceResponseHandler = (destResponse, sourceMessage, authKey, authori
...
+
+  switch (status) {
+    case 401:
+      // eslint-disable-next-line no-case-declarations
+      let errorCode = 'DEFAULT';
+      if (authKey && matchErrorCode('INVALID_SESSION_ID')) {
+        errorCode = 'INVALID_SESSION_ID';
+      }
+      handleAuthError(errorCode, authKey, authorizationFlow, sourceMessage, destResponse, status);
+      break;
```
Comment (sanpj2292) [inline src/v0/destinations/salesforce/utils.js:119]:
```suggestion
    case 401: {
        let errorCode = 'DEFAULT';
        if (authKey && matchErrorCode('INVALID_SESSION_ID')) {
          errorCode = 'INVALID_SESSION_ID';
        }
        handleAuthError(errorCode, authKey, authorizationFlow, sourceMessage, destResponse, status);
        break;
      }
```

Comment (sanpj2292) [reply src/v0/destinations/salesforce/utils.js:119]: Reference: https://eslint.org/docs/latest/rules/no-case-declarations

File: src/v0/destinations/salesforce/utils.js [hunk 7]
```diff
@@ -18,27 +15,42 @@ const {
...
+  }
+  return JSON.stringify(response);
+};
 
-  // if the response from destination is not a success case build an explicit error
-  if (!isHttpStatusSuccess(status) && status >= 400) {
-    const matchErrorCode = (errorCode) =>
-      response && Array.isArray(response) && response.some((resp) => resp?.errorCode === errorCode);
-    if (status === 401 && authKey && matchErrorCode('INVALID_SESSION_ID')) {
+const handleAuthError = (
```
Comment (sanpj2292) [inline src/v0/destinations/salesforce/utils.js:43]: I feel we can rewrite the function in a simpler way like below

```javascript
const handleAuthError = (
  errorCode,
  authKey,
  authorizationFlow,
  sourceMessage,
  destResponse,
  status,
) => {
  if (errorCode === 'INVALID_SESSION_ID') {
    let authErrCategory = '';
    if (authorizationFlow === OAUTH) {
      authErrCategory = getAuthErrCategoryFromStCode(status);
    } else {
      ACCESS_TOKEN_CACHE.del(authKey);
    }
    throw new RetryableError(
      `${DESTINATION} Request Failed - due to "INVALID_SESSION_ID", (${authErrCategory}) ${sourceMessage}`,
      500,
      destResponse,
      authErrCategory,
    );
  }

  throw new AbortedError(
    `${DESTINATION} Request Failed: "${status}" due to "${getErrorMessage(destResponse.response)}", (Aborted) ${sourceMessage}`,
    400,
    destResponse,
  );
};

```

**Comment by shrouti1507** [comment]
we are handling this in https://github.com/rudderlabs/rudder-transformer/pull/3824

**Pull request closed without merging**